File: slnSopimusrekisteri/appSopimusrekisteri/Controls/DateInput.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports AjaxControlToolkit

Namespace GeneralControls

  Public Class DateInput

    Inherits PlaceHolder
    Implements INamingContainer

    Private txtDate As TextBox
    Private imgPopup As Image
    Private calendar As CalendarExtender
    Private validaattori As CompareValidator

    Public Sub New()

      txtDate = New TextBox() With {.SkinID = "Datetime", .ID = "txtDate"}
      imgPopup = New Image() With {.SkinID = "CalendarImage", .AlternateText = "Valitse päivämäärä", .ID = "imgPopup"}
      validaattori = New CompareValidator() With {.ControlToValidate = txtDate.ID, .Operator = ValidationCompareOperator.DataTypeCheck, .Type = ValidationDataType.Date, .ErrorMessage = "Anna päivämäärä muodossa pp.mm.vvvv"}
      calendar = New CalendarExtender() With {.PopupButtonID = imgPopup.ID, .TargetControlID = txtDate.ID, .Format = "dd.MM.yyyy"}
    End Sub

    Protected Overrides Sub OnInit(e As EventArgs)
      MyBase.OnInit(e)

      Controls.Add(txtDate)
      Controls.Add(imgPopup)
      Controls.Add(calendar)
      Controls.Add(validaattori)
    End Sub

    Public Property Text As String
      Get
        Return txtDate.Text
      End Get
      Set
        txtDate.Text = Value
      End Set
    End Property

    Public Property DateValue As DateTime?
      Get
        If (String.IsNullOrEmpty(txtDate.Text)) Then
          Return Nothing
        End If
        Return DateTime.Parse(txtDate.Text)
      End Get
      Set
        If (Value.HasValue) Then
          txtDate.Text = Value.Value.ToShortDateString()
        End If
      End Set
    End Property

    Public Property ValidationGroup As String
      Get
        Return validaattori.ValidationGroup
      End Get
      Set
        validaattori.ValidationGroup = Value
      End Set
    End Property

    Public Property Enabled As Boolean
      Get
        Return txtDate.Enabled
      End Get
      Set
        txtDate.Enabled = Value
      End Set
    End Property
  End Class
End Namespace

File: Sopimusrekisteri.Controls/DateInput.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using AjaxControlToolkit;

namespace Sopimusrekisteri.Controls
{
  public class DateInput : PlaceHolder, INamingContainer
  {

    private TextBox txtDate;
    private Image imgPopup;
    private CalendarExtender calendar;

    public DateInput()
    {

    }

    protected override void OnInit(EventArgs e)
    {
      base.OnInit(e);

      this.txtDate = new TextBox() { SkinID = "Datetime" };
      this.imgPopup = new Image() { ImageUrl = this.ImageUrl, AlternateText = "Valitse päivämäärä" };
      this.calendar = new CalendarExtender();
      this.calendar.Format = "dd.MM.yyyy";

      this.Controls.Add(this.txtDate);
      this.Controls.Add(this.imgPopup);
      this.Controls.Add(this.calendar);

      this.txtDate.ID = "txtDate";
      this.imgPopup.ID = "imgPopup";
      calendar.PopupButtonID = this.imgPopup.ID;
      calendar.TargetControlID = this.txtDate.ID;
    }

    public string Text
    {
      get { return this.txtDate.Text; }
      set { this.txtDate.Text = value; }
    }

    public DateTime? DateValue
    {
      get
      {
        if (string.IsNullOrEmpty(this.txtDate.Text))
          return null;
        return DateTime.Parse(this.txtDate.Text);
      }
      set
      {
        if (value.HasValue)
          this.txtDate.Text = value.Value.ToShortDateString();
      }
    }

    public string ValidationGroup
    {
      get { return this.txtDate.ValidationGroup; }
      set { this.txtDate.ValidationGroup = value; }
    }

    public bool Enabled
    {
      get { return this.txtDate.Enabled; }
      set { this.txtDate.Enabled = value; }
    }

    public string ImageUrl { get; set; }

  }
}